src/ring_theory/unique_factorization_domain.lean

```diff
@@ -107,6 +107,17 @@ wf_dvd_monoid.induction_on_irreducible a
       by { rw multiset.prod_cons,
            exact hs.2.mul_left _ }⟩⟩)
 
+lemma exists_factors_eq (a : α) (hn0 : a ≠ 0) (hnu : ¬ is_unit a) :
+  ∃ (f : multiset α), (∀ (b : α), b ∈ f → irreducible b) ∧ f.prod = a :=
```

Comment:
How about we strengthen this to:
```lean
lemma not_unit_iff_exists_factors_eq (a : α) (hn0 : a ≠ 0) :
  ¬ is_unit a ↔ ∃ (f : multiset α), (∀ (b : α), b ∈ f → irreducible b) ∧ f.prod = a ∧ f ≠ ∅ :=
begin
  refine ⟨λ hnu, _, _⟩,
  { obtain ⟨P, hi, u, rfl⟩ := exists_factors a hn0,
    obtain ⟨p, h⟩ := multiset.exists_mem_of_ne_zero (λ h : P = 0, hnu $ by simp [h]),
    classical, use (P.erase p).cons (p * u), split,
    { intros a ha, rcases multiset.mem_cons.1 ha with (rfl|ha),
      exacts [associated.irreducible ⟨u,rfl⟩ (hi p h), hi a (multiset.mem_of_mem_erase ha)] },
    split,
    { rw [multiset.prod_cons, mul_comm p, mul_assoc, multiset.prod_erase h, mul_comm] },
    { simp, }, },
  { rintros ⟨f, hf, rfl, hne⟩,
    refine multiset.prod_induction_nonempty (λ a, ¬ is_unit a) (λ a b ha hb, _)
      hne (λ b hb, irreducible.not_unit (hf b hb)),
    contrapose! ha,
    obtain ⟨c, hc⟩ := ha.exists_right_inv,
    rw mul_assoc at hc,
    exact is_unit_iff_exists_inv.mpr ⟨_, hc⟩, },
end
```
(the proof could probably be tidied up)

Reply:
I think this is unlikely helpful, as the other direction only require much weaker hypothesis: a product with at least one non-unit factor is a non-unit. ([not_is_unit_of_not_is_unit_dvd](https://leanprover-community.github.io/mathlib_docs/algebra/divisibility.html#not_is_unit_of_not_is_unit_dvd), and irreducible -> non-unit by definition.)

Reply:
However, the f ≠ ∅ conclusion may be useful, and ↔ enables rewrite, so maybe we should do that.

Reply:
Would you prefer
```
  a ≠ 0 ∧ ¬ is_unit a ↔ ∃ f : multiset α, (∀ b ∈ f, irreducible b) ∧ f.prod = a :=
```
?